File: windows/KeyAction.cs
using System.Diagnostics;
using System.Runtime.InteropServices;
using System.Threading.Tasks;

namespace KeyboardDrop;

public enum KeyActionType
{
    Remap,
    App,
    Command,
    Ssh,
    File,
    Script,
    Screenshot,
    Url,
    Volume,
    Media,
    Text,
    System
}

public class KeyAction
{
    public KeyActionType Type { get; set; }
    public int RemapCode { get; set; }
    public string Target { get; set; } = "";

    public bool IsRemap => Type == KeyActionType.Remap;

    public string Description
    {
        get
        {
            return Type switch
            {
                KeyActionType.Remap => $"-> {KeyMap.NameFor(RemapCode)}",
                KeyActionType.App => $"-> App: {System.IO.Path.GetFileNameWithoutExtension(Target)}",
                KeyActionType.Command => $"-> Cmd: {(Target.Length > 30 ? Target[..27] + "..." : Target)}",
                KeyActionType.Ssh => $"-> SSH: {Target}",
                KeyActionType.File => $"-> File: {System.IO.Path.GetFileName(Target)}",
                KeyActionType.Script => $"-> Script: {System.IO.Path.GetFileName(Target)}",
                KeyActionType.Screenshot => $"-> Screenshot: {Target}",
                KeyActionType.Url => $"-> URL: {Target}",
                KeyActionType.Volume => $"-> Volume: {Target}",
                KeyActionType.Media => $"-> Media: {Target}",
                KeyActionType.Text => $"-> Text: {(Target.Length > 20 ? Target[..17] + "..." : Target)}",
                KeyActionType.System => $"-> System: {Target}",
                _ => "-> ???"
            };
        }
    }

    public void Execute()
    {
        switch (Type)
        {
            case KeyActionType.Remap:
                break;

            case KeyActionType.App:
                Process.Start(new ProcessStartInfo { FileName = Target, UseShellExecute = true });
                break;

            case KeyActionType.Command:
                Process.Start(new ProcessStartInfo
                {
                    FileName = "cmd.exe",
                    Arguments = $"/c {Target}",
                    UseShellExecute = false,
                    CreateNoWindow = true
                });
                break;

            case KeyActionType.Ssh:
                Process.Start(new ProcessStartInfo
                {
                    FileName = "cmd.exe",
                    Arguments = $"/k ssh {Target}",
                    UseShellExecute = true
                });
                break;

            case KeyActionType.File:
                Process.Start(new ProcessStartInfo { FileName = Target, UseShellExecute = true });
                break;

            case KeyActionType.Script:
                Process.Start(new ProcessStartInfo { FileName = Target, UseShellExecute = true });
                break;

            case KeyActionType.Screenshot:
                ExecuteScreenshot(Target);
                break;

            case KeyActionType.Url:
                Process.Start(new ProcessStartInfo { FileName = Target, UseShellExecute = true });
                break;

            case KeyActionType.Volume:
                ExecuteVolume(Target);
                break;

            case KeyActionType.Media:
                ExecuteMedia(Target);
                break;

            case KeyActionType.Text:
                Task.Run(() => ExecuteText(Target));
                break;

            case KeyActionType.System:
                ExecuteSystem(Target);
                break;
        }
    }

    private void ExecuteScreenshot(string mode)
    {
        var timestamp = System.DateTime.Now.ToString("yyyy-MM-dd_HH-mm-ss");
        var path = System.IO.Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.Desktop),
            $"Screenshot_{timestamp}.png");

        var args = mode.ToLowerInvariant() switch
        {
            "region" or "selection" => $"/c powershell -command \"Add-Type -AssemblyName System.Windows.Forms; $r = [System.Windows.Forms.Screen]::PrimaryScreen.Bounds; System.Drawing.Bitmap($r.Width, $r.Height)\"",
            "clipboard" or "clip" => "/c echo Screenshot to clipboard not implemented on Windows yet",
            _ => $"/c powershell -command \"Add-Type -AssemblyName System.Drawing; $screen = [System.Windows.Forms.Screen]::PrimaryScreen.Bounds; $bmp = New-Object System.Drawing.Bitmap($screen.Width, $screen.Height); $g = [System.Drawing.Graphics]::FromImage($bmp); $g.CopyFromScreen($screen.X, $screen.Y, 0, 0, $bmp.Size); $g.Dispose(); $bmp.Save('{path}')\""
        };
        Process.Start(new ProcessStartInfo
        {
            FileName = "cmd.exe",
            Arguments = args,
            UseShellExecute = false,
            CreateNoWindow = true
        });
    }

    private void ExecuteVolume(string action)
    {
        var args = action.ToLowerInvariant() switch
        {
            "up" => "/c nircmd.exe changesysvolume 5000",
            "down" => "/c nircmd.exe changesysvolume -5000",
            "mute" => "/c nircmd.exe mutesysvolume 1",
            "unmute" => "/c nircmd.exe mutesysvolume 0",
            _ => ""
        };
        if (!string.IsNullOrEmpty(args))
        {
            Process.Start(new ProcessStartInfo
            {
                FileName = "cmd.exe",
                Arguments = args,
                UseShellExecute = false,
                CreateNoWindow = true
            });
        }
    }

    private void ExecuteMedia(string action)
    {
        int vk = action.ToLowerInvariant() switch
        {
            "play" or "playpause" or "pause" => 0xB3,
            "next" or "next_track" => 0xB0,
            "prev" or "previous" or "prev_track" => 0xB1,
            "stop" => 0xB2,
            _ => -1
        };
        if (vk >= 0)
        {
            NativeMethods.keybd_event((byte)vk, 0, 0, System.IntPtr.Zero);
            NativeMethods.keybd_event((byte)vk, 0, NativeMethods.KEYEVENTF_KEYUP, System.IntPtr.Zero);
        }
    }

    private void ExecuteText(string text)
    {
        foreach (char c in text)
        {
            short vk = NativeMethods.VkKeyScan(c);
            if (vk != -1)
            {
                bool shift = (vk & 0x100) != 0;
                if (shift)
                    NativeMethods.keybd_event(0xA0, 0, 0, System.IntPtr.Zero); // shift down

                NativeMethods.keybd_event((byte)(vk & 0xFF), 0, 0, System.IntPtr.Zero);
                NativeMethods.keybd_event((byte)(vk & 0xFF), 0, NativeMethods.KEYEVENTF_KEYUP, System.IntPtr.Zero);

                if (shift)
                    NativeMethods.keybd_event(0xA0, 0, NativeMethods.KEYEVENTF_KEYUP, System.IntPtr.Zero); // shift up
            }
            System.Threading.Thread.Sleep(20);
        }
    }

    private void ExecuteSystem(string action)
    {
        var args = action.ToLowerInvariant() switch
        {
            "sleep" => "/c rundll32.exe powrprof.dll,SetSuspendState 0,1,0",
            "lock" => "/c rundll32.exe user32.dll,LockWorkStation",
            "restart" => "/c shutdown /r /t 0",
            "shutdown" => "/c shutdown /s /t 0",
            "logout" => "/c shutdown /l",
            _ => ""
        };
        if (!string.IsNullOrEmpty(args))
        {
            Process.Start(new ProcessStartInfo
            {
                FileName = "cmd.exe",
                Arguments = args,
                UseShellExecute = false,
                CreateNoWindow = true
            });
        }
    }

    public static KeyAction? FromConfig(object value)
    {
        if (value is string str)
        {
            var code = KeyMap.CodeFor(str);
            if (code == null)
            {
                System.Console.WriteLine($"  [warning] unknown target key: {str}");
                return null;
            }
            return new KeyAction { Type = KeyActionType.Remap, RemapCode = code.Value };
        }

        if (value is System.Text.Json.JsonElement el && el.ValueKind == System.Text.Json.JsonValueKind.Object)
        {
            string? type = null;
            string? target = null;
            foreach (var prop in el.EnumerateObject())
            {
                if (prop.Name == "type") type = prop.Value.GetString();
                if (prop.Name == "target") target = prop.Value.GetString();
            }

            if (type == null || target == null)
            {
                System.Console.WriteLine("  [warning] invalid action config");
                return null;
            }

            return type.ToLowerInvariant() switch
            {
                "remap" or "key" => KeyMap.CodeFor(target) is int c
                    ? new KeyAction { Type = KeyActionType.Remap, RemapCode = c }
                    : null,
                "app" => new KeyAction { Type = KeyActionType.App, Target = target },
                "command" or "cmd" => new KeyAction { Type = KeyActionType.Command, Target = target },
                "ssh" => new KeyAction { Type = KeyActionType.Ssh, Target = target },
                "file" => new KeyAction { Type = KeyActionType.File, Target = target },
                "script" => new KeyAction { Type = KeyActionType.Script, Target = target },
                "screenshot" or "screen" => new KeyAction { Type = KeyActionType.Screenshot, Target = target },
                "url" or "web" => new KeyAction { Type = KeyActionType.Url, Target = target },
                "volume" => new KeyAction { Type = KeyActionType.Volume, Target = target },
                "media" => new KeyAction { Type = KeyActionType.Media, Target = target },
                "text" or "type" => new KeyAction { Type = KeyActionType.Text, Target = target },
                "system" => new KeyAction { Type = KeyActionType.System, Target = target },
                _ => null
            };
        }

        return null;
    }
}

File: macos/KeyAction.swift
import Cocoa
import CoreGraphics

enum KeyAction: Equatable {
    case remap(CGKeyCode)
    case app(String)
    case command(String)
    case ssh(String)
    case file(String)
    case script(String)
    case screenshot(String)
    case url(String)
    case volume(String)
    case media(String)
    case text(String)
    case system(String)

    var isRemap: Bool {
        if case .remap = self { return true }
        return false
    }

    var description: String {
        switch self {
        case .remap(let code):
            let name = KeyMap.name(for: code) ?? String(format: "0x%02X", code)
            return "-> \(name)"
        case .app(let path):
            let name = (path as NSString).lastPathComponent.replacingOccurrences(of: ".app", with: "")
            return "-> App: \(name)"
        case .command(let cmd):
            let short = cmd.count > 30 ? String(cmd.prefix(27)) + "..." : cmd
            return "-> Cmd: \(short)"
        case .ssh(let host):
            return "-> SSH: \(host)"
        case .file(let path):
            return "-> File: \((path as NSString).lastPathComponent)"
        case .script(let path):
            return "-> Script: \((path as NSString).lastPathComponent)"
        case .screenshot(let mode):
            return "-> Screenshot: \(mode)"
        case .url(let u):
            return "-> URL: \(u)"
        case .volume(let action):
            return "-> Volume: \(action)"
        case .media(let action):
            return "-> Media: \(action)"
        case .text(let t):
            let short = t.count > 20 ? String(t.prefix(17)) + "..." : t
            return "-> Text: \(short)"
        case .system(let action):
            return "-> System: \(action)"
        }
    }

    func execute() {
        switch self {
        case .remap:
            break

        case .app(let path):
            NSWorkspace.shared.openApplication(at: URL(fileURLWithPath: path), configuration: .init())

        case .command(let cmd):
            runShell(cmd)

        case .ssh(let host):
            runAppleScript("tell application \"Terminal\"\nactivate\ndo script \"ssh \(host)\"\nend tell")

        case .file(let path):
            NSWorkspace.shared.open(URL(fileURLWithPath: path))

        case .script(let path):
            let task = Process()
            task.launchPath = "/bin/sh"
            task.arguments = [path]
            try? task.run()

        case .screenshot(let mode):
            let timestamp = formattedTimestamp()
            let desktop = NSHomeDirectory() + "/Desktop"
            let filename = "\(desktop)/Screenshot_\(timestamp).png"
            switch mode.lowercased() {
            case "region", "selection":
                runShell("screencapture -i '\(filename)'")
            case "window":
                runShell("screencapture -w '\(filename)'")
            case "clipboard", "clip":
                runShell("screencapture -c")
            default:
                runShell("screencapture '\(filename)'")
            }

        case .url(let urlString):
            if let url = URL(string: urlString) {
                NSWorkspace.shared.open(url)
            }

        case .volume(let action):
            switch action.lowercased() {
            case "up":
                runAppleScript("set volume output volume ((output volume of (get volume settings)) + 10)")
            case "down":
                runAppleScript("set volume output volume ((output volume of (get volume settings)) - 10)")
            case "mute":
                runAppleScript("set volume with output muted")
            case "unmute":
                runAppleScript("set volume without output muted")
            default:
                if let level = Int(action) {
                    runAppleScript("set volume \(max(0, min(100, level)))")
                }
            }

        case .media(let action):
            switch action.lowercased() {
            case "play", "playpause", "pause":
                runAppleScript("tell application \"Music\" to playpause")
            case "next", "next_track":
                runAppleScript("tell application \"Music\" to next track")
            case "prev", "previous", "prev_track":
                runAppleScript("tell application \"Music\" to previous track")
            default:
                break
            }

        case .text(let t):
            DispatchQueue.global(qos: .userInitiated).async {
                for char in t {
                    if let keyCode = charToKeyCode(char) {
                        let down = CGEvent(keyboardEventSource: nil, virtualKey: keyCode, keyDown: true)
                        let up = CGEvent(keyboardEventSource: nil, virtualKey: keyCode, keyDown: false)
                        down?.post(tap: .cghidEventTap)
                        up?.post(tap: .cghidEventTap)
                        usleep(20000)
                    }
                }
            }

        case .system(let action):
            switch action.lowercased() {
            case "sleep":
                runShell("pmset sleepnow")
            case "lock":
                runShell("/System/Library/CoreServices/Menu\\ Extras/User.menu/Contents/Resources/CGSession -suspend")
            case "restart":
                runAppleScript("tell application \"System Events\" to restart")
            case "shutdown":
                runAppleScript("tell application \"System Events\" to shut down")
            case "logout":
                runShell("kill -TERM $(pgrep -u $USER -x loginwindow)")
            default:
                break
            }
        }
    }

    static func fromConfig(_ value: Any) -> KeyAction? {
        if let str = value as? String {
            guard let code = KeyMap.code(for: str) else {
                print("  [warning] unknown target key: \(str)")
                return nil
            }
            return .remap(code)
        }

        guard let dict = value as? [String: Any],
              let type = dict["type"] as? String,
              let target = dict["target"] as? String else {
            print("  [warning] invalid action config, expected {\"type\": \"...\", \"target\": \"...\"}")
            return nil
        }

        switch type.lowercased() {
        case "remap", "key":
            guard let code = KeyMap.code(for: target) else {
                print("  [warning] unknown target key: \(target)")
                return nil
            }
            return .remap(code)
        case "app":       return .app(target)
        case "command", "cmd": return .command(target)
        case "ssh":       return .ssh(target)
        case "file":      return .file(target)
        case "script":    return .script(target)
        case "screenshot", "screen": return .screenshot(target)
        case "url", "web": return .url(target)
        case "volume":    return .volume(target)
        case "media":     return .media(target)
        case "text", "type": return .text(target)
        case "system":    return .system(target)
        default:
            print("  [warning] unknown action type: \(type)")
            return nil
        }
    }
}

// MARK: - Helpers

private func runShell(_ cmd: String) {
    let task = Process()
    task.launchPath = "/bin/sh"
    task.arguments = ["-c", cmd]
    try? task.run()
}

private func runAppleScript(_ source: String) {
    if let script = NSAppleScript(source: source) {
        var error: NSDictionary?
        script.executeAndReturnError(&error)
    }
}

private func formattedTimestamp() -> String {
    let f = DateFormatter()
    f.dateFormat = "yyyy-MM-dd_HH-mm-ss"
    return f.string(from: Date())
}

private func charToKeyCode(_ char: Character) -> CGKeyCode? {
    let lower = String(char).lowercased()
    guard let first = lower.first else { return nil }
    return KeyMap.code(for: String(first))
}
